auth/src/test/java/com/firebase/ui/auth/testhelpers/TestHelper.java

```diff
@@ -15,44 +15,105 @@
 
 import android.content.Context;
+import android.content.res.Resources;
 
 import com.firebase.ui.auth.AuthUI;
 import com.firebase.ui.auth.AuthUI.IdpConfig;
 import com.firebase.ui.auth.IdpResponse;
+import com.firebase.ui.auth.R;
 import com.firebase.ui.auth.data.model.FlowParameters;
 import com.google.firebase.FirebaseApp;
 import com.google.firebase.FirebaseOptions;
+import com.google.firebase.auth.EmailAuthProvider;
+import com.google.firebase.auth.FacebookAuthProvider;
+import com.google.firebase.auth.FirebaseAuth;
 import com.google.firebase.auth.FirebaseUser;
+import com.google.firebase.auth.GoogleAuthProvider;
+import com.google.firebase.auth.PhoneAuthProvider;
+import com.google.firebase.auth.TwitterAuthProvider;
 
 import org.mockito.ArgumentCaptor;
+import org.robolectric.RuntimeEnvironment;
 
+import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 
 import static junit.framework.Assert.assertEquals;
 import static junit.framework.Assert.assertNotNull;
 import static junit.framework.Assert.assertNull;
+import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 public class TestHelper {
     private static final String APPLICATION_ID = "testAppId";
     private static final String API_KEY = "fakeKey";
-    private static final String FIREBASE_APP_NAME = "firebaseAppName";
 
-    public static FirebaseApp initializeApp(Context context) {
+    public static void initialize() {
+        initializeProviders();
+        AuthUI.setApplicationContext(RuntimeEnvironment.application);
+        FirebaseApp app = initializeApp(RuntimeEnvironment.application);
+        mockFirebaseAuth(app);
+    }
+
+    private static FirebaseApp initializeApp(Context context) {
         try {
             return FirebaseApp.initializeApp(
                     context,
                     new FirebaseOptions.Builder()
                             .setApiKey(API_KEY)
                             .setApplicationId(APPLICATION_ID)
                             .build(),
-                    FIREBASE_APP_NAME);
+                    FirebaseApp.DEFAULT_APP_NAME);
         } catch (IllegalStateException e) {
-            return FirebaseApp.getInstance(FIREBASE_APP_NAME);
+            return FirebaseApp.getInstance(FirebaseApp.DEFAULT_APP_NAME);
+        }
+    }
+
+    private static void mockFirebaseAuth(FirebaseApp app) {
+        for (Field field : FirebaseAuth.class.getDeclaredFields()) {
+            field.setAccessible(true);
+
+            Object o;
+            try {
+                o = field.get(null);
+            } catch (IllegalAccessException e) {
+                throw new IllegalStateException(e);
+            } catch (NullPointerException e) {
+                continue;
+            }
+
+            if (o instanceof Map) {
```

Comment:
What is going on here ... I need some more comments and probably a justification for this extreme hackery haha

Reply:
Whoops, yeah. This came about due to my frustration at not being able to properly spy/mock the auth instance. The method basically injects a mock FirebaseAuth into the instances map. I've added some documentation.

In case you were wondering, it's what lets us get rid of `AuthUI.getTestInstance`. I prefer clean production code and dirty tests than dirty prod code and clean tests. Hope that's cool with you! 😄